compiler/rustc_passes/src/stability.rs

```diff
@@ -805,14 +806,69 @@ impl<'tcx> Visitor<'tcx> for Checker<'tcx> {
     }
 
     fn visit_path(&mut self, path: &'tcx hir::Path<'tcx>, id: hir::HirId) {
-        if let Some(def_id) = path.res.opt_def_id() {
-            let method_span = path.segments.last().map(|s| s.ident.span);
-            self.tcx.check_stability(def_id, Some(id), path.span, method_span)
+        if !is_unstable_reexport(self.tcx, path, id) {
+            if let Some(def_id) = path.res.opt_def_id() {
+                let method_span = path.segments.last().map(|s| s.ident.span);
+                self.tcx.check_stability(def_id, Some(id), path.span, method_span)
+            }
         }
         intravisit::walk_path(self, path)
     }
 }
 
+/// Check whether the path is a re-export of an unstable item, that has also been marked as unstable under the same feature name.
+fn is_unstable_reexport<'a, 'tcx>(
+    tcx: TyCtxt<'tcx>,
+    path: &'a hir::Path<'tcx>,
+    id: hir::HirId,
+) -> bool {
+    // Get the LocalDefId so we can lookup the item to check the kind.
+    let def_id = if let Some(def_id) = tcx.hir().opt_local_def_id(id) {
+        def_id
+    } else {
+        return false;
+    };
+
+    let stab = if let Some(stab) = tcx.stability().local_stability(def_id) {
+        stab
+    } else {
+        return false;
+    };
+
+    let item = tcx.hir().item(hir::ItemId { def_id });
+    let item_kind = &item.kind;
+
+    // If this is a path that isn't a use, we don't need to do anything special
+    if !matches!(item_kind, ItemKind::Use(..)) {
+        return false;
+    }
+
+    let decision = if let Some(def_id) = path.res.opt_def_id() {
+        let method_span = path.segments.last().map(|s| s.ident.span);
+
+        tcx.eval_stability(def_id, Some(id), path.span, method_span)
+    } else {
+        // If the path doesn't resolve to anything, there's nothing more to check.
+        return false;
+    };
+
+    // If we are going to deny the usage, check whether it's a re-export first with the same unstable feature.
+    if let rustc_middle::middle::stability::EvalResult::Deny { feature, .. } = decision {
+        if stab.level.is_unstable() && feature == stab.feature {
+            // This is a re-export, marked as unstable, with the same feature name. Don't require a `#![feature]` gate
+            return true;
+        } else {
+            // We almost met the requirements but the feature name didn't match. Emit a warning in addition to the standard feature gate error.
+            tcx.sess.parse_sess.struct_warn("Re-exporting an unstable item without a feature gate requires marking the re-export as unstable with the same feature name")
+                // should probably be getting the span of the unstable attr here? Maybe make it an actual suggestion.
+                .span_help(path.span, format!("Use `#![unstable(feature = {})]` to re-export", feature))
+                .emit();
```

Comment:
Hard-coded warnings should almost never be reported, and this warning is not generally useful.
If the reexport is gated by a separate feature then there are probably reasons for that.